### PR TITLE
[MetricsAdvisor] Added Event Hubs and Log Analytics data feed sources

### PR DESCRIPTION
**Pull request by kinelski**
Part of https://github.com/azure/azure-sdk-for-net/issues/18610.

### CODE REVIEW

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs [hunk 1]
```diff
@@ -740,9 +746,9 @@ public abstract partial class DataFeedSource
         public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType AzureDataExplorer { get { throw null; } }
         public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType AzureDataLakeStorageGen2 { get { throw null; } }
         public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType AzureEventHubs { get { throw null; } }
-        public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType AzureLogAnalytics { get { throw null; } }
         public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType AzureTable { get { throw null; } }
         public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType InfluxDb { get { throw null; } }
+        public static Azure.AI.MetricsAdvisor.Models.DataFeedSourceType LogAnalytics { get { throw null; } }
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs:751]: Need to double check with the team why we remove the `Azure` prefix here. Other `Azure` data sources don't have the same change.

Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs:751]: https://github.com/Azure/azure-sdk-for-net/issues/21625

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs [hunk 2]
```diff
@@ -24,14 +24,16 @@ dataFeedDetail switch
                 AzureApplicationInsightsDataFeed d => new AzureApplicationInsightsDataFeedSource(d.DataSourceParameter),
                 AzureBlobDataFeed d => new AzureBlobDataFeedSource(d.DataSourceParameter),
                 AzureCosmosDBDataFeed d => new AzureCosmosDbDataFeedSource(d.DataSourceParameter),
+                AzureDataExplorerDataFeed d => new AzureDataExplorerDataFeedSource(d.DataSourceParameter),
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs:27]: Moved this and MongoDB to keep them in alphabetical order.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs [hunk 3]
```diff
@@ -24,14 +24,16 @@ dataFeedDetail switch
                 AzureApplicationInsightsDataFeed d => new AzureApplicationInsightsDataFeedSource(d.DataSourceParameter),
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSource.cs:24]: This file contains the transforms we do from public type to service type.
- Service => Public (used in Get, List, and the return of the Update operation).
- Public => Service (used in Create)
- Public => Service Patch (used in Update)

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/tests/MetricsAdvisorAdministrationClient/DataFeedLiveTests.cs [hunk 4]
```diff
@@ -239,6 +244,48 @@ public async Task CreateAndGetAzureDataLakeStorageGen2DataFeedWithOptionalMember
             ValidateAzureDataLakeStorageGen2DataSource(createdDataFeed.DataSource as AzureDataLakeStorageGen2DataFeedSource);
         }
 
+        [RecordedTest]
+        [Ignore("https://github.com/Azure/azure-sdk-for-net/issues/21623")]
+        public async Task CreateAndGetAzureEventHubsDataFeedWithMinimumSetup()
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/tests/MetricsAdvisorAdministrationClient/DataFeedLiveTests.cs:249]: Two issues here (both tracked by https://github.com/Azure/azure-sdk-for-net/issues/21623).
- The service does not accept an Event Hubs data source with "mock" parameters, as we do for other data sources. We need to set up an actual Event Hubs resource to test this.
- I ran the tests locally and they pass except for one property (`IngestionStartTime`) that the service seems to be storing incorrectly. Double checked and it does not seem to be on client-side. I'll follow up with them.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/tests/Models/DataFeedSourcesTests.cs [hunk 5]
```diff
@@ -24,9 +24,11 @@ public class DataFeedSourcesTests : MockClientTestBase
             new object[] { new AzureCosmosDbDataFeedSource("secret", "mock", "mock", "mock"), "connectionString" },
             new object[] { new AzureDataExplorerDataFeedSource("secret", "mock"), "connectionString" },
             new object[] { new AzureDataLakeStorageGen2DataFeedSource("mock", "secret", "mock", "mock", "mock"), "accountKey" },
+            new object[] { new AzureEventHubsDataFeedSource("secret", "mock"), "connectionString" },
```
Comment (kinelski) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/tests/Models/DataFeedSourcesTests.cs:27]: I loved the fact that the changes to the non-live tests were minimum. I'll think about if it's possible to do the same for our live tests, but only when I have time. Tried to think a bit, but there are more scenarios.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs [hunk 6]
```diff
@@ -920,6 +926,16 @@ public partial class InfluxDbDataFeedSource : Azure.AI.MetricsAdvisor.Models.Dat
         public static bool operator !=(Azure.AI.MetricsAdvisor.Models.IngestionStatusType left, Azure.AI.MetricsAdvisor.Models.IngestionStatusType right) { throw null; }
         public override string ToString() { throw null; }
     }
+    public partial class LogAnalyticsDataFeedSource : Azure.AI.MetricsAdvisor.Models.DataFeedSource
+    {
+        public LogAnalyticsDataFeedSource(string workspaceId, string query) { }
```
Comment (maririos) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs:931]: so a secret is not needed for all cases?

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        /// <summary>
+        /// The tenant ID used for AAD authentication.
+        /// </summary>
+        public string TenantId { get; set; }
+
+        /// <summary>
+        /// The client secret used for AAD authentication.
+        /// </summary>
+        internal string ClientSecret { get; set; }
```
Comment (maririos) [inline sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs:92]: should have the `Volatile` stuff to make sure it is thread-safe, right?

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs (hunk 6 above)
Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/api/Azure.AI.MetricsAdvisor.netstandard2.0.cs:931]: Correct. For example, a data source may have public read-only access enabled, making authentication unnecessary. I need to revisit some data sources and check with the service which authentication properties are required and which are not (https://github.com/Azure/azure-sdk-for-net/issues/21616).

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs (hunk 7 above)
Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs:92]: You're right. Totally missed it. Updating it.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    public class LogAnalyticsDataFeedSource : DataFeedSource
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LogAnalyticsDataFeedSource"/> class.
+        /// </summary>
+        /// <param name="workspaceId">The workspace ID of the Log Analytics resource.</param>
+        /// <param name="query">The query to retrieve the data to be ingested.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="workspaceId"/> or <paramref name="query"/> is null.</exception>
+        /// <exception cref="ArgumentException"><paramref name="workspaceId"/> or <paramref name="query"/> is empty.</exception>
+        public LogAnalyticsDataFeedSource(string workspaceId, string query)
```
[the remark this answers is not in the document]
Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs:21]: In this case there's no authentication specified. Replied in the other conversation.

File: sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs (hunk 7 above)
Comment (kinelski) [reply sdk/metricsadvisor/Azure.AI.MetricsAdvisor/src/Models/DataFeed/DataFeedSources/LogAnalyticsDataFeedSource.cs]: Done.

**Review by maririos — APPROVED** [review]

**Pull request merged**